### PR TITLE
Add basic Client model with appropriate attributes

### LINKED ISSUE #48: Client Model

### LINKED ISSUE #73: Tests for serializing client

### PR DESCRIPTION
**Pull request by Douglch**
Closes #48, closes #73.

## Changes done so far
- [x] `Client` class, adding the necessary attributes and getters.
- [x] `ClientNotFoundException` and `DuplicateClientException` for error handling.
- [x] `JsonAdaptedClient` class to convert the Jackson-friendly adapted client object into the model's `Client` object.
- [x] Client 1/2 tests
- [x] Client 2/2 tests
- [x] Client List
- [x] Client CRUD methods

### Note
Will be implementing more test cases to pass codecov. Will also implement Source class, all to be done on Sun.

### CODE REVIEW

File: src/main/java/mycelium/mycelium/model/client/YearOfBirth.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package mycelium.mycelium.model.client;
+
+import static mycelium.mycelium.commons.util.AppUtil.checkArgument;
+
+/**
+ * The YearOfBirth class represents a year of birth in the application.
+ * It provides methods for validating a year of birth, and storing and retrieving a year of birth value.
+ */
+public class YearOfBirth {
```
Comment (immanuelhume) [inline src/main/java/mycelium/mycelium/model/client/YearOfBirth.java:9]: Did `java.time.Year` have some issues?

File: src/main/java/mycelium/mycelium/storage/JsonAdaptedClient.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        final Email modelEmail = new Email(email);
+
+        nullCheck(yearOfBirth == null, YearOfBirth.class.getSimpleName());
+        // validityCheck(!YearOfBirth.isValidYearOfBirth(yearOfBirth), YearOfBirth.MESSAGE_CONSTRAINTS);
+        final YearOfBirth modelYearOfBirth = new YearOfBirth(yearOfBirth);
+
+        // TODO validityCheck for source
+        nullCheck(source == null, String.format(MISSING_FIELD_MESSAGE_FORMAT, "source"));
+
+        nullCheck(mobileNumber == null, Phone.class.getSimpleName());
```
Comment (immanuelhume) [inline src/main/java/mycelium/mycelium/storage/JsonAdaptedClient.java:104]: These three fields can actually be null in JSON, so we must skip the null checks for them.

Comment (immanuelhume) [reply src/main/java/mycelium/mycelium/storage/JsonAdaptedClient.java:104]: And do the validitycheck only if they are not null

File: src/test/java/mycelium/mycelium/model/client/YearOfBirthTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Test
+    public void constructor_invalidPhone_throwsIllegalArgumentException() {
+        String invalidYearOfBirth = "";
+        assertThrows(IllegalArgumentException.class, () -> new YearOfBirth(invalidYearOfBirth));
+    }
+
+    @Test
+    public void isValidYearOfBirth() {
```
Comment (immanuelhume) [inline src/test/java/mycelium/mycelium/model/client/YearOfBirthTest.java:22]: nice tests :+1:

File: src/test/java/mycelium/mycelium/storage/JsonAdaptedClientTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import org.junit.jupiter.api.Test;
+
+import mycelium.mycelium.commons.exceptions.IllegalValueException;
+import mycelium.mycelium.model.client.YearOfBirth;
+import mycelium.mycelium.model.person.Email;
+import mycelium.mycelium.model.person.Name;
+
+
+
+public class JsonAdaptedClientTest {
```
Comment (immanuelhume) [inline src/test/java/mycelium/mycelium/storage/JsonAdaptedClientTest.java:16]: I think we should also test the serializing of the entire Client, i.e. create a client, serialize, then read it back and assert that we get the same client back. I'll just add to issue, can do next week.

**Review by immanuelhume — CHANGES_REQUESTED** [review]
LGTM! Just a small change required in JsonAdaptedClient

File: src/main/java/mycelium/mycelium/model/client/YearOfBirth.java (hunk 1 above)
Comment (Douglch) [reply src/main/java/mycelium/mycelium/model/client/YearOfBirth.java:9]: I was using it initially, but found out that Year's a static class and monad, where the of() method only accepts int, so I had to convert them somewhere in the code from String to int. But the main reason why I created the `YearOfBirth` class was coz I wanted to create the validity check, and I could just make everything stay as a String.

File: src/test/java/mycelium/mycelium/model/client/YearOfBirthTest.java (hunk 3 above)
Comment (Douglch) [reply src/test/java/mycelium/mycelium/model/client/YearOfBirthTest.java:22]: thx 🥵

File: src/main/java/mycelium/mycelium/storage/JsonAdaptedClient.java (hunk 2 above)
Comment (Douglch) [reply src/main/java/mycelium/mycelium/storage/JsonAdaptedClient.java]: done 👍

File: src/main/java/mycelium/mycelium/storage/JsonSerializableAddressBook.java [hunk 5]
```diff
@@ -21,9 +22,11 @@
 class JsonSerializableAddressBook {
 
     public static final String MESSAGE_DUPLICATE_PERSON = "Persons list contains duplicate person(s).";
+    public static final String MESSAGE_DUPLICATE_CLIENT = "Persons list contains duplicate client(s).";
```
Comment (Douglch) [inline src/main/java/mycelium/mycelium/storage/JsonSerializableAddressBook.java:25]: To be edited

**Review by immanuelhume — APPROVED** [review]
:shipit:

**Pull request merged**